### ISSUE TITLE
Gradle doesn't lock used configuration when build is run

### ISSUE THREAD

**Issue by xenoterracide** [gradle/gradle#28534]
### Current Behavior

gradle isn't locking my `errorprone` configuration when I run `build` even though `errorprone` is run when I run build. My understanding is that when running a task that with `--write-locks` dependencies of that task should be written. Thus running `build --write-locks` since `errorprone` hooks into classes (or something) as an annotation processor and uses the `errorprone``configuration` the errorprone configuration itself should be locked.

### Expected Behavior

errorprone should be locked

### Context (optional)

I don't want my dependencies to written when used

### Steps to Reproduce

I am an incredibly lazy human being, I'm imagining this alone will do it, then run `./gradlew build --write-locks`, no locks with the configuration scope `errorprone` will be written. run `./gradlew dependencies --write-locks` this will add `errorprone` to the locks.

```kts
buildscript { dependencyLocking { lockAllConfigurations() } }

plugins {
    `java-library`
    id("net.ltgt.errorprone") version "3.+"
}
repositories {
    mavenCentral()
}
dependencyLocking {
  lockAllConfigurations()
}
dependencies {
    errorprone("com.google.errorprone:error_prone_core:2.+")
}
```

In the event this simple build script doesn't do it, it might have something to do with using `buildSrc`, full code can be found https://github.com/xenoterracide/java-commons/blob/ae2d9bfa8c9fca5e094200c55eb5706cca12c75e/buildSrc/src/main/kotlin/our.javacompile.gradle.kts#L16

### Gradle version

8.6

### Your Environment (optional)

```
------------------------------------------------------------
Gradle 8.6
------------------------------------------------------------

Build time:   2024-02-02 16:47:16 UTC
Revision:     d55c486870a0dc6f6278f53d21381396d0741c6e

Kotlin:       1.9.20
Groovy:       3.0.17
Ant:          Apache Ant(TM) version 1.10.13 compiled on January 4 2023
JVM:          21.0.2 (Eclipse Adoptium 21.0.2+13-LTS)
OS:           Linux 6.6.19-1-MANJARO amd64
```

**Comment by ov7a** [comment]
Most probably your issue is caused by an external Gradle plugin, which cannot be addressed in the Gradle Build Tool project. We cannot just move the issue.

We recommend you open an issue on its issue tracker.

We have some integration tests with major plugins to catch potential problems early, but we don't have the capacity to cover all plugins available.

If you think you have come across a Gradle bug, please provide a reproducer for this issue without the plugin, and we’ll consider reopening it.

***

This can be explained by the following. The `errorprone` plugin does not _resolve_ the `errorprone` configuration. If that's the case, the configuration should not be declared as resolvable. To verify that, do you have an entry for `error_prone_core` in the lock file generated with build?

**Comment by tbroyer** [comment]
While I agree that the `errorprone` configuration should not be resolvable (it is for backwards compatibility mostly), and changing it would solve the "problem" for that one specific project, there's no guarantee that `build` will run every task and resolve every configuration. And actually this is also the case with the `errorproneJavac` configuration: it's only used (and thus resolved) if compiling with a JDK 8 toolchain, so with any other version of the JDK it won't be resolved during a `build` but will be during `dependencies`.

In other words: it's not really that the errorprone plugin should change, it's more about expectations of the OP that aren't aligned with how Gradle works.

**Comment by xenoterracide** [comment]
More like the OP doesn't agree with how gradle works... I dislike playing guessing games on how to lock my dependencies. Currently that's how great it works. I'm on my phone and don't feel like locating the bug that basically says we want a single command to actually do this. Currently though, it's guessing games.
